File: sdks/java/src/test/java/io/t402/client/DiscoveryClientTest.java
package io.t402.client;

import com.sun.net.httpserver.HttpServer;
import io.t402.model.DiscoveryModels;
import io.t402.model.DiscoveryModels.*;
import io.t402.util.Json;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class DiscoveryClientTest {

    private HttpServer server;
    private DiscoveryClient client;
    private int port;

    @BeforeEach
    void setUp() throws IOException {
        server = HttpServer.create(new InetSocketAddress(0), 0);
        port = server.getAddress().getPort();
        client = new DiscoveryClient("http://localhost:" + port);
    }

    @AfterEach
    void tearDown() {
        if (server != null) {
            server.stop(0);
        }
    }

    @Test
    void listResources() throws Exception {
        DiscoveryResponse mockResp = new DiscoveryResponse();
        mockResp.t402Version = 2;
        DiscoveryItem item = new DiscoveryItem();
        item.id = "res-1";
        item.resource = "https://example.com/api";
        item.type = "http";
        item.t402Version = 2;
        item.lastUpdated = 1700000000L;
        item.accepts = List.of();
        mockResp.items = List.of(item);
        PaginationInfo page = new PaginationInfo();
        page.limit = 20;
        page.offset = 0;
        page.total = 1;
        mockResp.pagination = page;

        server.createContext("/discovery/resources", exchange -> {
            assertEquals("GET", exchange.getRequestMethod());
            String query = exchange.getRequestURI().getQuery();
            assertNotNull(query);
            assertTrue(query.contains("type=http"));
            assertTrue(query.contains("limit=10"));

            byte[] body = Json.MAPPER.writeValueAsBytes(mockResp);
            exchange.sendResponseHeaders(200, body.length);
            try (OutputStream os = exchange.getResponseBody()) {
                os.write(body);
            }
        });
        server.start();

        ListResourcesParams params = new ListResourcesParams();
        params.type = "http";
        params.limit = 10;

        DiscoveryResponse resp = client.listResources(params);
        assertEquals(2, resp.t402Version);
        assertEquals(1, resp.items.size());
        assertEquals("res-1", resp.items.get(0).id);
        assertEquals(1, resp.pagination.total);
    }

    @Test
    void listResourcesNullParams() throws Exception {
        DiscoveryResponse mockResp = new DiscoveryResponse();
        mockResp.t402Version = 2;
        mockResp.items = List.of();
        PaginationInfo page = new PaginationInfo();
        page.limit = 20;
        page.offset = 0;
        page.total = 0;
        mockResp.pagination = page;

        server.createContext("/discovery/resources", exchange -> {
            assertNull(exchange.getRequestURI().getQuery());
            byte[] body = Json.MAPPER.writeValueAsBytes(mockResp);
            exchange.sendResponseHeaders(200, body.length);
            try (OutputStream os = exchange.getResponseBody()) {
                os.write(body);
            }
        });
        server.start();

        DiscoveryResponse resp = client.listResources(null);
        assertEquals(0, resp.items.size());
    }

    @Test
    void getResource() throws Exception {
        DiscoveryItem mockItem = new DiscoveryItem();
        mockItem.id = "res-123";
        mockItem.resource = "https://example.com/api";
        mockItem.type = "http";
        mockItem.t402Version = 2;
        mockItem.lastUpdated = 1700000000L;
        PaymentOption opt = new PaymentOption("exact", "eip155:8453", "100", "USDT", "0xabc", 3600);
        mockItem.accepts = List.of(opt);

        server.createContext("/discovery/resources/res-123", exchange -> {
            assertEquals("GET", exchange.getRequestMethod());
            byte[] body = Json.MAPPER.writeValueAsBytes(mockItem);
            exchange.sendResponseHeaders(200, body.length);
            try (OutputStream os = exchange.getResponseBody()) {
                os.write(body);
            }
        });
        server.start();

        DiscoveryItem item = client.getResource("res-123");
        assertEquals("res-123", item.id);
        assertEquals(1, item.accepts.size());
        assertEquals("exact", item.accepts.get(0).scheme);
    }

    @Test
    void getResourceNotFound() throws Exception {
        server.createContext("/discovery/resources/nonexistent", exchange -> {
            byte[] body = "{\"error\":\"not found\"}".getBytes();
            exchange.sendResponseHeaders(404, body.length);
            try (OutputStream os = exchange.getResponseBody()) {
                os.write(body);
            }
        });
        server.start();

        IOException ex = assertThrows(IOException.class,
                () -> client.getResource("nonexistent"));
        assertTrue(ex.getMessage().contains("404"));
    }

    @Test
    void registerResource() throws Exception {
        RegisterResourceResponse mockResp = new RegisterResourceResponse();
        mockResp.id = "res-new";
        mockResp.resource = "https://example.com/api";
        mockResp.type = "http";
        mockResp.t402Version = 2;
        mockResp.createdAt = "2026-01-15T10:00:00Z";

        server.createContext("/discovery/register", exchange -> {
            assertEquals("POST", exchange.getRequestMethod());

            String reqBody = new String(exchange.getRequestBody().readAllBytes());
            RegisterResourceRequest parsed =
                    Json.MAPPER.readValue(reqBody, RegisterResourceRequest.class);
            assertEquals("https://example.com/api", parsed.resource);
            assertEquals("http", parsed.type);
            assertEquals(1, parsed.accepts.size());

            byte[] body = Json.MAPPER.writeValueAsBytes(mockResp);
            exchange.sendResponseHeaders(201, body.length);
            try (OutputStream os = exchange.getResponseBody()) {
                os.write(body);
            }
        });
        server.start();

        RegisterResourceRequest req = new RegisterResourceRequest(
                "https://example.com/api", "http",
                List.of(new PaymentOption("exact", "eip155:8453", "100", "USDT", "0xabc", 3600))
        );
        RegisterResourceResponse resp = client.registerResource(req);
        assertEquals("res-new", resp.id);
    }

    @Test
    void registerResourceConflict() throws Exception {
        server.createContext("/discovery/register", exchange -> {
            byte[] body = "{\"error\":\"already exists\"}".getBytes();
            exchange.sendResponseHeaders(409, body.length);
            try (OutputStream os = exchange.getResponseBody()) {
                os.write(body);
            }
        });
        server.start();

        RegisterResourceRequest req = new RegisterResourceRequest(
                "https://example.com/api", "http",
                List.of(new PaymentOption("exact", "eip155:8453", "100", "USDT", "0xabc", 3600))
        );
        IOException ex = assertThrows(IOException.class,
                () -> client.registerResource(req));
        assertTrue(ex.getMessage().contains("409"));
    }

    @Test
    void updateResource() throws Exception {
        DiscoveryItem mockItem = new DiscoveryItem();
        mockItem.id = "res-123";
        mockItem.resource = "https://example.com/api";
        mockItem.type = "http";
        mockItem.t402Version = 2;
        mockItem.lastUpdated = 1700000000L;
        mockItem.accepts = List.of();

        server.createContext("/discovery/resources/res-123", exchange -> {
            assertEquals("PUT", exchange.getRequestMethod());

            String reqBody = new String(exchange.getRequestBody().readAllBytes());
            UpdateResourceRequest parsed =
                    Json.MAPPER.readValue(reqBody, UpdateResourceRequest.class);
            assertTrue(parsed.active);

            byte[] body = Json.MAPPER.writeValueAsBytes(mockItem);
            exchange.sendResponseHeaders(200, body.length);
            try (OutputStream os = exchange.getResponseBody()) {
                os.write(body);
            }
        });
        server.start();

        UpdateResourceRequest req = new UpdateResourceRequest();
        req.active = true;
        DiscoveryItem item = client.updateResource("res-123", req);
        assertEquals("res-123", item.id);
    }

    @Test
    void deleteResource() throws Exception {
        server.createContext("/discovery/resources/res-123", exchange -> {
            assertEquals("DELETE", exchange.getRequestMethod());
            exchange.sendResponseHeaders(204, -1);
        });
        server.start();

        assertDoesNotThrow(() -> client.deleteResource("res-123"));
    }

    @Test
    void deleteResourceForbidden() throws Exception {
        server.createContext("/discovery/resources/res-123", exchange -> {
            byte[] body = "{\"error\":\"not authorized\"}".getBytes();
            exchange.sendResponseHeaders(403, body.length);
            try (OutputStream os = exchange.getResponseBody()) {
                os.write(body);
            }
        });
        server.start();

        IOException ex = assertThrows(IOException.class,
                () -> client.deleteResource("res-123"));
        assertTrue(ex.getMessage().contains("403"));
    }
}

File: sdks/go/http/discovery_client_test.go
package http

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func TestListResources(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "GET" {
			t.Errorf("expected GET, got %s", r.Method)
		}
		if r.URL.Path != "/v1/discovery/resources" {
			t.Errorf("expected /discovery/resources, got %s", r.URL.Path)
		}
		if r.URL.Query().Get("type") != "http" {
			t.Errorf("expected type=http, got %s", r.URL.Query().Get("type"))
		}
		if r.URL.Query().Get("limit") != "10" {
			t.Errorf("expected limit=10, got %s", r.URL.Query().Get("limit"))
		}
		_ = json.NewEncoder(w).Encode(DiscoveryResponse{
			T402Version: 2,
			Items: []DiscoveryItem{
				{ID: "res-1", Resource: "https://example.com/api", Type: "http", T402Version: 2},
			},
			Pagination: PaginationInfo{Limit: 10, Offset: 0, Total: 1},
		})
	}))
	defer ts.Close()

	client := NewHTTPFacilitatorClient(&FacilitatorConfig{URL: ts.URL})
	resp, err := client.ListResources(context.Background(), &ListResourcesParams{Type: "http", Limit: 10})
	if err != nil {
		t.Fatalf("ListResources failed: %v", err)
	}
	if len(resp.Items) != 1 {
		t.Errorf("expected 1 item, got %d", len(resp.Items))
	}
	if resp.Items[0].ID != "res-1" {
		t.Errorf("expected id res-1, got %s", resp.Items[0].ID)
	}
	if resp.Pagination.Total != 1 {
		t.Errorf("expected total 1, got %d", resp.Pagination.Total)
	}
}

func TestListResourcesNilParams(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.RawQuery != "" {
			t.Errorf("expected no query params, got %s", r.URL.RawQuery)
		}
		_ = json.NewEncoder(w).Encode(DiscoveryResponse{
			T402Version: 2,
			Items:       []DiscoveryItem{},
			Pagination:  PaginationInfo{Limit: 20, Offset: 0, Total: 0},
		})
	}))
	defer ts.Close()

	client := NewHTTPFacilitatorClient(&FacilitatorConfig{URL: ts.URL})
	resp, err := client.ListResources(context.Background(), nil)
	if err != nil {
		t.Fatalf("ListResources failed: %v", err)
	}
	if len(resp.Items) != 0 {
		t.Errorf("expected 0 items, got %d", len(resp.Items))
	}
}

func TestGetResource(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "GET" {
			t.Errorf("expected GET, got %s", r.Method)
		}
		if r.URL.Path != "/v1/discovery/resources/res-123" {
			t.Errorf("expected /discovery/resources/res-123, got %s", r.URL.Path)
		}
		_ = json.NewEncoder(w).Encode(DiscoveryItem{
			ID:          "res-123",
			Resource:    "https://example.com/api",
			Type:        "http",
			T402Version: 2,
			Accepts: []DiscoveryPaymentOption{
				{Scheme: "exact", Network: "eip155:8453", Amount: "100", Asset: "USDT", PayTo: "0xabc"},
			},
			LastUpdated: time.Now().Unix(),
		})
	}))
	defer ts.Close()

	client := NewHTTPFacilitatorClient(&FacilitatorConfig{URL: ts.URL})
	item, err := client.GetResource(context.Background(), "res-123")
	if err != nil {
		t.Fatalf("GetResource failed: %v", err)
	}
	if item.ID != "res-123" {
		t.Errorf("expected id res-123, got %s", item.ID)
	}
	if len(item.Accepts) != 1 {
		t.Errorf("expected 1 payment option, got %d", len(item.Accepts))
	}
}

func TestGetResourceNotFound(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusNotFound)
		_, _ = w.Write([]byte(`{"error":"not found"}`))
	}))
	defer ts.Close()

	client := NewHTTPFacilitatorClient(&FacilitatorConfig{URL: ts.URL})
	_, err := client.GetResource(context.Background(), "nonexistent")
	if err == nil {
		t.Fatal("expected error for not found resource")
	}
}

func TestRegisterResource(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "POST" {
			t.Errorf("expected POST, got %s", r.Method)
		}
		if r.URL.Path != "/v1/discovery/register" {
			t.Errorf("expected /discovery/register, got %s", r.URL.Path)
		}

		var req RegisterResourceRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			t.Fatalf("failed to decode request body: %v", err)
		}
		if req.Resource != "https://example.com/api" {
			t.Errorf("expected resource https://example.com/api, got %s", req.Resource)
		}
		if req.Type != "http" {
			t.Errorf("expected type http, got %s", req.Type)
		}
		if len(req.Accepts) != 1 {
			t.Errorf("expected 1 payment option, got %d", len(req.Accepts))
		}

		w.WriteHeader(http.StatusCreated)
		_ = json.NewEncoder(w).Encode(RegisterResourceResponse{
			ID:          "res-new",
			Resource:    req.Resource,
			Type:        req.Type,
			T402Version: 2,
			CreatedAt:   time.Now(),
		})
	}))
	defer ts.Close()

	client := NewHTTPFacilitatorClient(&FacilitatorConfig{URL: ts.URL})
	resp, err := client.RegisterResource(context.Background(), &RegisterResourceRequest{
		Resource: "https://example.com/api",
		Type:     "http",
		Accepts: []DiscoveryPaymentOption{
			{Scheme: "exact", Network: "eip155:8453", Amount: "100", Asset: "USDT", PayTo: "0xabc", MaxTimeoutSeconds: 3600},
		},
	})
	if err != nil {
		t.Fatalf("RegisterResource failed: %v", err)
	}
	if resp.ID != "res-new" {
		t.Errorf("expected id res-new, got %s", resp.ID)
	}
}

func TestUpdateResource(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "PUT" {
			t.Errorf("expected PUT, got %s", r.Method)
		}
		if r.URL.Path != "/v1/discovery/resources/res-123" {
			t.Errorf("expected /discovery/resources/res-123, got %s", r.URL.Path)
		}

		var req UpdateResourceRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			t.Fatalf("failed to decode request body: %v", err)
		}
		if req.Active == nil || !*req.Active {
			t.Error("expected active=true")
		}

		_ = json.NewEncoder(w).Encode(DiscoveryItem{
			ID:          "res-123",
			Resource:    "https://example.com/api",
			Type:        "http",
			T402Version: 2,
		})
	}))
	defer ts.Close()

	client := NewHTTPFacilitatorClient(&FacilitatorConfig{URL: ts.URL})
	active := true
	item, err := client.UpdateResource(context.Background(), "res-123", &UpdateResourceRequest{Active: &active})
	if err != nil {
		t.Fatalf("UpdateResource failed: %v", err)
	}
	if item.ID != "res-123" {
		t.Errorf("expected id res-123, got %s", item.ID)
	}
}

func TestDeleteResource(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != "DELETE" {
			t.Errorf("expected DELETE, got %s", r.Method)
		}
		if r.URL.Path != "/v1/discovery/resources/res-123" {
			t.Errorf("expected /discovery/resources/res-123, got %s", r.URL.Path)
		}
		w.WriteHeader(http.StatusNoContent)
	}))
	defer ts.Close()

	client := NewHTTPFacilitatorClient(&FacilitatorConfig{URL: ts.URL})
	err := client.DeleteResource(context.Background(), "res-123")
	if err != nil {
		t.Fatalf("DeleteResource failed: %v", err)
	}
}

func TestDeleteResourceForbidden(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusForbidden)
		_, _ = w.Write([]byte(`{"error":"not authorized"}`))
	}))
	defer ts.Close()

	client := NewHTTPFacilitatorClient(&FacilitatorConfig{URL: ts.URL})
	err := client.DeleteResource(context.Background(), "res-123")
	if err == nil {
		t.Fatal("expected error for forbidden delete")
	}
}

func TestListResourcesWithAllParams(t *testing.T) {
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		q := r.URL.Query()
		if q.Get("network") != "eip155:8453" {
			t.Errorf("expected network eip155:8453, got %s", q.Get("network"))
		}
		if q.Get("scheme") != "exact" {
			t.Errorf("expected scheme exact, got %s", q.Get("scheme"))
		}
		if q.Get("category") != "ai" {
			t.Errorf("expected category ai, got %s", q.Get("category"))
		}
		_ = json.NewEncoder(w).Encode(DiscoveryResponse{
			T402Version: 2,
			Items:       []DiscoveryItem{},
			Pagination:  PaginationInfo{Limit: 20, Offset: 0, Total: 0},
		})
	}))
	defer ts.Close()

	client := NewHTTPFacilitatorClient(&FacilitatorConfig{URL: ts.URL})
	_, err := client.ListResources(context.Background(), &ListResourcesParams{
		Network:  "eip155:8453",
		Scheme:   "exact",
		Category: "ai",
	})
	if err != nil {
		t.Fatalf("ListResources failed: %v", err)
	}
}
